### PR TITLE
Allow migration generator to parse select key-value options from CLI

### PR DESCRIPTION
**Pull request by fractaledmind**
### Motivation / Background

The current generators that create migration files (`migration`, `model`, `resource`, `scaffold`) allow a small subset of column options to be passed from the command line:
* You can pass a `limit` to `string`, `text`, `binary`, or `integer` fields using an integer within curly braces (e.g. `name:string{255}`).
* You can pass `precision` and `scale` to `decimal` fields using two integers within curly braces (e.g. `latitude:decimal{15,10}`).
* You can also pass the `polymorphic` option to `belongs_to` or `references` fields (e.g. `commentable:references{polymorphic}`).

There are, however, various other column options that can't be passed in via the command line, like `null` or `default`.

In my opinion, Rails' generators are one of its key features. And I'm confident we have all written a generator command that we _knew_ was going to require us to open up the migration file _just_ to add `null: false` to one or two fields. Frustrating and unnecessary. Let's get make the migration generator powerful.

### Detail

This PR powers up the column definition string passed from the CLI to contain key-value options, both for columns and for indexes. Like the existing single keyword options, the key-value options are passed within curly braces (`{}`) using a format like `column_name:type{key:value}`. As with precision and scale options, multiple options can be passed in if separated by a comma, like so `column_name:type{key:value,foo:bar}`. As seen in the new `AVAILABLE_COLUMN_OPTIONS ` constant, the allows keys are `%w(limit default null precision scale array polymorphic foreign_key if_exists)`, which matches the options documented across the various migration methods for fields. Acceptable values include `true`, `false`, `[]`, `{}`, and any quote-wrapped string or integer or float. Moreover, this PR allows column definitions to include spaces so that the CLI test can be more easily scanned.

Tests provide some examples:
* `title:string{null:false}`
* `description:text{default: 'hello, world!'}`
* `tags:text{array, null:false, default:[]}`
* `owner:references{foreign_key:{table_name:users}}`

As said, users would also now be able to pass options for indexes as well. Key-value options accept the same range of values, and only a specified set of keys are permitted: `%w(unique length order opclass where type using comment algorithm name if_not_exists internal polymorphic)` (as set in the `AVAILABLE_INDEX_OPTIONS` constant). This allows for column definitions like:

* `title:string{null: false}:uniq{name: by_title}`
* `title:string:index{length:{title:15}}`

Altogether, these additions (column options, index options, spaces allowed) make many generators immediately more powerful.

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

### CODE REVIEW

**Comment by fractaledmind** [comment]
I should add, I am completely open to adding more tests for more migration types. I just wanted to get the basic idea in front of maintainers first before investing more time into the tests. If you like the feature, then I'll fill out the test suite to cover `create_table` and `remove_column`

**Comment by andrewculver** [comment]
Having the frequently used options available on the CLI seems like a win. As it is right now, when you're writing a tutorial or recording a screencast, you end up with two steps, one for `rails g` and then another instructing them on how to go into the migration file and which changes to make. Would love to be able to skip the second step more often.

I also wouldn't mind being able to specify `foreign_key: {table_name: ...}` from the CLI somehow. That one comes up a lot for me when working with related models in the same namespace.

**Comment by fractaledmind** [comment]
@andrewculver: Good point. I didn't want to mutate this PR until I get some more feedback, but after some refactoring, I was able to get the type and options parser to handle nested key-value pairs as well. I have a PR against my fork demonstrating the code and the initial test case: https://github.com/fractaledmind/rails/pull/1

This would enable specifying `owner:references{foreign_key={table_name=users}}` from the CLI

**Comment by fractaledmind** [comment]
Also, in conversation with @kaspth, I explored using `:` as the separator for the key-value pairs in addition to being the top-level separator between name, type, and index_type. That was made much easier by building on the refactoring in the referenced PR to enable nested key-value pairs. The tweak on top of that code can be found in this PR: https://github.com/fractaledmind/rails/pull/2

Again, I'm happy to merge these in and fill out this PR as soon as someone signals they like that direction.

**Comment by fractaledmind** [comment]
@andrewculver: Curious on your thoughts about the choice between `=` and `:` for the key-value separator. What @kaspth and I like about `:` is that the CLI text looks close to a Ruby hash. In the most complex case, the difference would be:

* `owner:references{foreign_key:{table_name:users},null:false}:uniq`
* `owner:references{foreign_key={table_name=users},null=false}:uniq`

**Comment by erlingur** [comment]
Personally I would prefer using `:` because it's also what ends up in the migration file. I'm pretty sure I would always instinctively write `:` and then have to remember it's supposed to be `=`.

**Comment by erlingur** [comment]
Also, this is a great PR! I don't know how often I've googled how to pass `null: false` to a generator just to find out you can't. Thanks! I really hope this gets merged.

File: railties/lib/rails/generators/generated_attribute.rb [hunk 1]
```diff
@@ -88,6 +96,29 @@ def parse_type_and_options(type)
...
+                      when "true" then true
+                      when "false" then false
+                      when "[]" then []
+                      when "{}" then {}
+                      when /^['"].*['"]$/ then val[1...-1]
+                      when nil then true
+                      when /\A\d+\z/ then Integer(val)
+                      else val
+                end
+                [key.to_sym, val]
```
Comment (revskill10) [inline railties/lib/rails/generators/generated_attribute.rb:118]: Can we just use `key.to_sym, val` here (without []) ?

Comment (fractaledmind) [reply railties/lib/rails/generators/generated_attribute.rb:118]: I don't believe so. I believe you only have that option when using a keyword like `return`

Comment (revskill10) [reply railties/lib/rails/generators/generated_attribute.rb]: @fractaledmind You're correct. So in this case, should we use `return key.to_sym, val` as i guess it's standard Ruby for multiple return value ?

**Comment by fractaledmind** [comment]
I don't think Rails has appetite for this. Closing.

**Comment by erlingur** [comment]
That's sad to hear, I was looking forward to this improvement 😕

**Comment by fractaledmind** [comment]
Can put it in a gem.

**Comment by erlingur** [comment]
That would actually be a pretty nice gem I think! I would add it for sure.

**Comment by jagthedrummer** [comment]
It's disappointing that such a useful and well written PR couldn't get any traction. I'd definitely use it if it were a gem. Or who do we have to petition to get it considered for rails itself?

**Comment by rafaelfranca** [comment]
I have no problem on giving the same answer here.

Thank you for the pull request. We don't want to complicate the generators and migration helpers to support all possible cases. They are targeted for the most common cases. For everything else, Rails can do, but you need to do a little of the work yourself.

**Comment by indigotechtutorials** [comment]
does this effect the old code in any way? I just don't see the down side to adding in a helpful improvement like this

**Pull request closed without merging**